### PR TITLE
Allocate tuples retrieved by a SharedScan in a longer living memory context

### PR DESCRIPTION
**Pull request by armenatzoglou**
When SharedScan requests a tuple from the underlying Sort, and the Sort has spilled to disk, the mk_sort code (also the regular sort) makes a copy of the memtuple, but places it in the sortcontext.
When hitting a QueryFinishPending (e.g., master received enough tuples, so stop the execution), EagerFree is called for the entire tree. When reaching the sort node, the sortcontext is deleted, even though the SharedScan above it still has a pointer to a memtuple allocated in this context.

The solution is:
pass a memory context for the mk_sort and regular sort to use (change the API and implementation of the following two functions: tuplesort_gettupleslot_pos_mk and tuplesort_gettupleslot_pos). Those functions should use the context provided by the caller to make a copy of the memtuple.

@foyzur @hlinnaka Please take a look if you have time. Thanks

### CODE REVIEW

File: src/test/regress/sql/query_finish_pending.sql [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+--start_ignore
+\! gpfaultinjector -f execsort_mksort_mergeruns -y reset --seg_dbid 2
+-- set QueryFinishPending=true in sort mergeruns. This will stop sort and set result_tape to NULL
+\! gpfaultinjector -f execsort_mksort_mergeruns -y finish_pending --seg_dbid 2
+--end_ignore
+
+-- return results although sort will be interrupted in one of the segments 
+select DISTINCT S from (select row_number() over(partition by i1 order by i2) AS T, count(*) over (partition by i1) AS S from _tmp_table) AS TMP;
+
+-- test if shared input scan deletes memory correctly when QuerFinishPending and its child has been eagerly freed,
```
Comment (danielgustafsson) [inline src/test/regress/sql/query_finish_pending.sql]: s/QuerFinishPending/QueryFinishPending/

File: src/test/regress/sql/query_finish_pending.sql [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+-- This will eagerly free the memory context of shared input scan's child node.  
+\! gpfaultinjector -f execshare_input_next -y finish_pending --seg_dbid 2
+--end_ignore
+
+select COUNT(i2) over(partition by i1)
+from testsisc
+LIMIT 2;
+
+
+-- test if shared input scan deletes memory correctly when QuerFinishPending and its child has been eagerly freed,
```
Comment (danielgustafsson) [inline src/test/regress/sql/query_finish_pending.sql]: Same as above.

**Comment by danielgustafsson** [comment]
Seems reasonable to me.

File: src/backend/executor/nodeShareInputScan.c [hunk 3]
```diff
@@ -213,11 +211,11 @@ ShareInputNext(ShareInputScanState *node)
 		{
 			if(gp_enable_mk_sort)
 			{
-				gotOK = tuplesort_gettupleslot_pos_mk(node->ts_state->sortstore_mk, (TuplesortPos_mk *)node->ts_pos, forward, slot);
+				gotOK = tuplesort_gettupleslot_pos_mk(node->ts_state->sortstore_mk, (TuplesortPos_mk *)node->ts_pos, forward, slot, CurrentMemoryContext);
```
Comment (foyzur) [inline src/backend/executor/nodeShareInputScan.c]: Can we at least assert about the expectation of CurrentMemoryContext?

File: src/backend/utils/sort/tuplesort_mk.c [hunk 4]
```diff
@@ -1198,7 +1203,7 @@ grow_unsorted_array(Tuplesortstate_mk *state)
 static void
 puttuple_common(Tuplesortstate_mk *state, MKEntry *e)
 {
-    Assert(is_under_sort_ctxt(state));
+    Assert(is_under_sort_or_exec_ctxt(state));
```
Comment (foyzur) [inline src/backend/utils/sort/tuplesort_mk.c]: I see, the assert is here.

**Comment by foyzur** [comment]
LGTM.

File: src/backend/executor/nodeShareInputScan.c (hunk 3 above)
Comment (armenatzoglou) [inline src/backend/executor/nodeShareInputScan.c]: Good point! I will do.

**Comment by armenatzoglou** [comment]
Pushed c36ecd6f6fafa64840aac793590988ecfe329751. Thanks a lot for your review.

**Pull request closed without merging**